### PR TITLE
Use AsyncBatchingWorkQueue instead of TaskQueue

### PR DESCRIPTION
**Pull request by sharwell**
This change was made during a review of [AB#1949588](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1949588). During the review, I failed to recognize a red herring in the automated analysis pipeline, which suggested that a high-hitting bug in 17.8 was still present in 17.9 and newer. In reality, the issue was already fixed for 17.9 Preview 1 by #70039, and internal analysis data confirms that there are no remaining known cases where `DiagnosticService._eventQueue` is consuming significant amounts of memory.

### CODE REVIEW

**Comment by Pilchie** [comment]
✅ Successfully linked to Azure Boards work item(s):
- [Bug 1949588](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1949588)

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs [hunk 1]
```diff
@@ -47,7 +48,11 @@ internal partial class DiagnosticService : IDiagnosticService
             _updateSources = [];
 
             // queue to serialize events.
-            _eventQueue = new TaskQueue(listenerProvider.GetListener(FeatureAttribute.DiagnosticService), TaskScheduler.Default);
+            _eventQueue = new AsyncBatchingWorkQueue<(IDiagnosticUpdateSource source, ImmutableArray<DiagnosticsUpdatedArgs> argsCollection)>(
+                delay: TimeSpan.Zero,
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs:52]: i'm wary about this.  we don't really do a zero timespan anywhere.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs:52]: is it possible ti imporve the IThreadingContext stuff so we can use the Dipose cancellation token for this guy?

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs [hunk 2]
```diff
@@ -65,51 +70,69 @@ internal partial class DiagnosticService : IDiagnosticService
...
+        private void RaiseDiagnosticsUpdated(IDiagnosticUpdateSource source, ImmutableArray<DiagnosticsUpdatedArgs> argsCollection)
         {
-            var ev = _eventMap.GetEventHandlers<EventHandler<ImmutableArray<DiagnosticsUpdatedArgs>>>(DiagnosticsUpdatedEventName);
-
-            _eventQueue.ScheduleTask(DiagnosticsUpdatedEventName, () =>
+            if (argsCollection.IsEmpty)
             {
-                using var argsBuilder = TemporaryArray<DiagnosticsUpdatedArgs>.Empty;
+                // No work to do. Avoid adding an empty args collection to _eventQueue because that is used to represent
+                // an operation from RaiseDiagnosticsCleared.
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs:116]: is this the same logic we had preivously?  trying to find it.

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs [hunk 3]
```diff
@@ -65,51 +70,69 @@ internal partial class DiagnosticService : IDiagnosticService
...
+            }
+
+            _eventQueue.AddWork((source, argsCollection));
+        }
 
-                // don't create event listener if it haven't created yet. if there is a diagnostic to remove
-                // listener should have already created since all events are done in the serialized queue
-                ev.RaiseEvent(static (handler, arg) => handler(arg.source, arg.args), (source, args: argsBuilder.ToImmutableAndClear()));
-            }, CancellationToken.None);
+        private void RaiseDiagnosticsCleared(IDiagnosticUpdateSource source)
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs:133]: i'm not really understanding who is calling this.

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs:116]: ok. i think i've wrapped my head around this.  but i def find it confusing.  could we instead enqueue this information as a boolean into the ABWQ instead?

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs [hunk 4]
```diff
@@ -65,51 +70,69 @@ internal partial class DiagnosticService : IDiagnosticService
...
+        {
+            _eventQueue.AddWork((source, ImmutableArray<DiagnosticsUpdatedArgs>.Empty));
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs:135]: i think i get it now.  but i'd prefer a bool passed along in this tuple to say if it's a clear or not.  the inference and managing of empty arrays is confusing IMO.

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs (hunk 1 above)
Comment (sharwell) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs:52]: `IThreadingContext` is not available in this layer, but we can do exactly the same thing by just making this MEF part disposable.

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs (hunk 2 above)
Comment (sharwell) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs:116]: ⏱️ Switching to an enum

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs (hunk 4 above)
Comment (sharwell) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs:135]: ➡️ Resolving this as part of https://github.com/dotnet/roslyn/pull/72425#discussion_r1515233241

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs (hunk 3 above)
Comment (sharwell) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs:133]: The caller is abstracted a few layers. There is an event hookup here:

https://github.com/dotnet/roslyn/blob/3ab3ca743a0a702a1dd925c40a969dc9c660dd5b/src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService_UpdateSourceRegistrationService.cs#L24

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs (hunk 1 above)
Comment (sharwell) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs:53]: ➡️ Updated

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs (hunk 2 above)
Comment (sharwell) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs]: ➡️ Updated

**Comment by sharwell** [comment]
@CyrusNajmabadi any other feedback here?

**Pull request merged**